java/23ai-jdbc-vector-search/src/main/java/com/oracle/dev/jdbc/OracleAIVectorSearchWithJava.java

```diff
@@ -0,0 +1,173 @@
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.Properties;
+import java.util.concurrent.ThreadLocalRandom;
+
+import oracle.jdbc.OracleType;
+import oracle.ucp.jdbc.PoolDataSource;
+import oracle.ucp.jdbc.PoolDataSourceFactory;
+
+public class OracleAIVectorSearchWithJava {
+
+	private final static String URL = "<JDBC_CONNECTION_URL>";
+	private final static String USERNAME = "<ORACLE_DATABASE_USERNAME>";
+	private final static String PASSWORD = "<ORACLE_DATABASE_PASSWORD>";
+
+	private String insertSql = "INSERT INTO ORACLE_AI_VECTOR_SEARCH_DEMO (ID, VECTOR_DATA) VALUES (?, ?)";
+	private String querySql = "SELECT ID, VECTOR_DATA FROM ORACLE_AI_VECTOR_SEARCH_DEMO";
+	private String querySqlWithBind = "SELECT ID, VECTOR_DATA FROM ORACLE_AI_VECTOR_SEARCH_DEMO ORDER BY VECTOR_DISTANCE(VECTOR_DATA, ?, DOT)";
+
+	public static void main(String[] args) throws SQLException {
+		OracleAIVectorSearchWithJava oracleAIVectorSearch = new OracleAIVectorSearchWithJava();
+		oracleAIVectorSearch.execute();
+	}
+
+	private void execute() throws SQLException {
+
+		System.out.println("Starting JDBC connection with PooledDataSource...");
+		try (Connection conn = getConnectionFromPooledDataSource()) {
+			System.out.println("Connected to Oracle Database 23ai! " + "\n");
+
+			System.out.println("Inserting VECTOR with oracle.jdbc.OracleType.VECTOR...");
+			insertVector(conn);
+			System.out.println("VECTOR with oracle.jdbc.OracleType.VECTOR inserted!" + "\n");
+
+			System.out.println("Inserting VECTOR with oracle.jdbc.OracleType.VARCHAR2...");
+			insertVectorWithVarChar2(conn);
+			System.out.println("VECTOR with oracle.jdbc.OracleType.VARCHAR2 inserted!" + "\n");
+
+			System.out.println("Inserting VECTOR with Batch API...");
+			insertVectorWithBatchAPI(conn);
+			System.out.println("VECTOR with Batch API inserted!" + "\n");
+
+			System.out.println("Retrieving VECTOR as double array...");
+			retrieveVectorAsArray(conn);
+			System.out.println("VECTOR retrieved!" + "\n");
+
+			System.out.println("Retrieving VECTOR as String...");
+			retrieveVectorAsString(conn);
+			System.out.println("VECTOR retrieved!" + "\n");
+
+			System.out.println("Retrieving VECTOR with bound VECTOR (L2 - Euclidean distance)...");
+			retrieveVectorWithBoundVector(conn);
+			System.out.println("VECTOR retrieved!");
+
+		}
+	}
+
+	private void insertVector(Connection connection) throws SQLException {
+		PreparedStatement insertStatement = connection.prepareStatement(insertSql);
+		double[] vector = { 1.1, 2.2, 3.3 };
+		System.out.println("SQL DML: " + insertSql);
+		System.out.println("VECTOR to be inserted: " + Arrays.toString(vector));
+		insertStatement.setInt(1, randomize());
+		insertStatement.setObject(2, vector, OracleType.VECTOR);
+		insertStatement.executeUpdate();
+	}
+
+	private void insertVectorWithVarChar2(Connection connection) throws SQLException {
+		PreparedStatement insertStatement = connection.prepareStatement(insertSql);
```

Comment:
Please use try with resources in such methods as this code as-is leaks cursors (the statement isn't closed).